documentation/docs/05-misc/02-testing.md

```diff
@@ -2,7 +2,156 @@
+
+### Using runes inside your test files
+
+It is possible to use runes inside your test files. First ensure your bundler knows to route the file through the Svelte compiler before running the test by adding `.svelte` to the filename (e.g `multiplier.svelte.test.js`). After that, you can use runes inside your tests.
+
+```js
+/// file: counter.svelte.test.js
+import { flushSync, mount, unmount } from 'svelte';
+import { expect, test } from 'vitest';
+import { multiplier } from './multiplier.svelte.js';
+
+test('Multiplier', () => {
+	let count = $state(0);
+	let double = multiplier(() => count, 2);
+
+	expect(double.value).toEqual(0);
+
+	count = 5;
+
+	expect(double.value).toEqual(10);
+});
+```
+
+If the tested code uses effects, you need to wrap the test inside `$effect.root` to create a scope in which these are properly captured.
+
+```js
+/// file: logger.svelte.test.js
+import { flushSync, mount, unmount } from 'svelte';
+import { expect, test } from 'vitest';
+import { logger } from './logger.svelte.js';
+
+test('Effect', () => {
+	const cleanup = $effect.root(() => {
+		let count = $state(0);
+		// logger uses an $effect to log updates of its input
+		let log = logger(() => count);
+
+		// effects normally run after a microtask,
+		// use flushSync to execute all pending effects synchronously
+		flushSync();
+		expect(log.value).toEqual([0]);
+
+		count = 1;
+		flushSync();
+
+		expect(log.value).toEqual([0, 1]);
+	});
+
+	cleanup();
+});
+```
+
+## E2E tests using Playwright
+
+E2E (short for "end to end") tests allow you to test your full application through the eyes of the user. This section uses [Playwright](https://playwright.dev/) as an example, but you can also use other solutions like [Cypress](https://www.cypress.io/) or [NightwatchJS](https://nightwatchjs.org/).
+
+To get start with Playwright, either let you guide by [their VS Code extension](https://playwright.dev/docs/getting-started-vscode), or install it from the command line using `npm init playwright`. It is also part of the setup CLI when you run `npm create svelte`.
+
+After you've done that, you should have a `tests` folder and a playwright config. You may need to adjust that config to tell Playwright what to do before running the tests - mainly starting your application at a certain port:
+
+```js
+/// file: playwright.config.js
+const config = {
+	webServer: {
+		command: 'npm run build && npm run preview',
+		port: 4173
+	},
+	testDir: 'tests',
+	testMatch: /(.+\.)?(test|spec)\.[jt]s/
+};
+
+export default config;
+```
+
+You can now start writing tests. These are totally unaware of Svelte as a framework, so you mainly interact with the DOM and write assertions.
+
+```js
+/// file: tests/hello-world.spec.js
+import { expect, test } from '@playwright/test';
+
+test('home page has expected h1', async ({ page }) => {
+	await page.goto('/');
```

Comment:
This is probably too much from a brief introduction but i had to copy the sveltekit implementation of the playwright fixture to allow navigations to wait for hydration.

Maybe a huge `details` that hide all of that but at least mention the problem?

Reply:
Using `page.locator` and similar APIs shouldn't make that a requirement because they automatically poll? We're not using that inside our test setup to have more strict tests around the exact timing at which things appear or disappear to test SvelteKit behavior itself - that shouldn't be necessary for user apps.

Reply:
> Using `page.locator` and similar APIs shouldn't make that a requirement because they automatically poll? We're not using that inside our test setup to have more strict tests around the exact timing at which things appear or disappear to test SvelteKit behavior itself - that shouldn't be necessary for user apps.

Oh actually it will probably not be a problem in svelte 5 but we had problems when plywright is too quick to write in an input and hydration deletes the value....but now hydration should preserve those.